### ISSUE TITLE
Error inconsistence

### ISSUE THREAD

**Issue by treziac** [confluentinc/confluent-kafka-dotnet#203]
In Produce, we throw error when Produce return non 0 (librdkafka know it won't be produced),complete and fill the Error property of Message when it's a non immediate error
So in the case the topic is unknown, we will have an unthrown exception sometimes (a message with the error field not null, because it's not flagged as permanent unknown topic), then throw an exception - for the same call with the same argument
This seems shaky in the .NET client.
What about never throwing in the Produce and instead add a HandleException in IDeliveryHandler, which would be call instead of throwing? Produce would never throw in this case, the only errors from producer would come from OnError (and/or OnPoduceError if we add this- but that's also the problem between which one we chose)
I think there is many place now where we can get error, and we should document them more - in the message, OnConsumeError, OnError, throwing when producing...

I would go with this:

OnError: error in the internal library. Not bound to producing or consuming a specific message. Broker unavailable, network failure... The error ismore or less severe and could be only informative (link to a post in librdkafka)

OnConsumeError: failed to consume a message. This may or may not be specific to a message (in which case topic, partition and offset will be set accordingly in Message). See the Message.Error for more details

ProduceAsync: will not throw because of Confluent.Kafka (may because of async .NET framework) The delivery report may contain error, in which case it will be passed through IDeliveryHandler.HandleDeliveryReport(Message) with Error field set or IDeliveryHandler.HandleException(Error)

**Comment by mhowlett** [comment]
yep - this needs some thought, but I suspect we can do better here.

We could possibly just call `HandleDeliveryReport` on the `IDeliveryHandler` immediately with the error value and other info we have that was supplied to the produce method. 

Initial reaction to `OnProduceError` is not into that idea.

**Comment by edenhill** [comment]
Would it make sense to sort this out in librdkafka instead?
We could add a configuration property that propagates these types of error (or all, QUEUE_FULL too?) through the delivery report too.

**Comment by treziac** [comment]
Could make sense. But I see the point in librdkafka to return directly rather than using deliveryreportcallback for immediate failure - let you see if it's better to have a single error endpoint or not in C, we can easily handle it on our side, just that it's message in .NET client to have so many errors endpoints and behaviour (throw vs not throw...)

I'm wondering by the way, why does librdkafka return RD_KAFKA_RESP_ERR__UNKNOWN_TOPIC marked as permanent state in the doc? If we create the topic after calling rd_kafka_produce, won't it be non-permanent?

**Comment by edenhill** [comment]
Temporary errors are those that stand a chance of resolving "automatically" within reasonable time:
connection errors, temporary states (no leader for partition), etc.

Permanent errors are the opposite, those that will require manual intervention:
 if a topic does not exist and is not auto-created then it'll require some external entity (such as a human) to create it, we'll not hold our breath waiting for that to happen.

**Comment by treziac** [comment]
Understood, thanks :)

**Comment by treziac** [comment]
Well, preventing ProduceAsync to throw is an issue when user don't provide an IDeliveryHandler as we can't inform him - but not providing IDeliveryHandler can also be considered as a "fire and forget", so this seems ok.

As this is related, why was `disableDeliveryReport `introduced ?
If we enable it and use the Task ProduceAsync, then the task is not awaitable, so there is a problem ( #201 )
If we pass null to void ProduceAsync, we achieve the same things than enabling this config (and as it returns void, it's more logical)

=> I would remove it and let the user pass null or not, or use the appropriote config. 
- Non-zero return would be treated as either:
  - a Task exception, called by IDeliveryHandler.HandleException
  - a DeliveryHandler with Message.Error setted (I would go for this one)
- Task ProduceAsync would directly set an exception (via one of the both option above) to inform there will be no callback if callbacks are disabled

This way:
- No more throw
- return Task always complete, eventually with Error

**Comment by mhowlett** [comment]
`disableDeliveryReport` was added as an optimization. You are right - there is absolutely no need for it as the code currently stands (I'm guessing at one point i wasn't checking for `deliveryHandler != null`. We should remove this.
